### PR TITLE
Main tedge config refactoring

### PR DESCRIPTION
**Pull request by jarhodes314**
## Proposed changes

*TLDR for people that are interested in the changes but don't want to review masses of code*: almost all the changes here arise from `new_tedge_config.rs`, run `cargo doc -p tedge-config --open` to see what the result of that is, and the screenshots below demonstrate the main user-facing changes.

This is the main chunk of work associated with #1812 (sorry it's so enormous). There are still a few things to be done in tedge_config, as well as public facing documentation about the updates.

If a user uses a key with lots of dots rather than the updated key, using tedge config will look something like this (I also renamed software.plugin.default -> software.default_plugin as the word order became particularly clunky after replacing the dot with an underscore):

image

This is what `tedge config list --doc` now outputs (assuming it's running in a tty, if stdout is piped elsewhere, the colours are disabled):

tedge config list --doc

All of the information shown here is derived from the doc comments, examples and notes in the various DTO structs. The colours and alignment are completely configurable, I just chose something that looked readable to me.

I haven't attempted yet to modify any crates other than tedge config and tedge.

I've attempted to add doc comments and examples for the major parts I've added. It's probably worth looking in the generated documentation (`cargo doc -p tedge_config --open` will get you there quickly if you want to try that), although I imagine people don't tend to run `cargo doc` an awful lot.

I also fixed some unrelated tests so they work on my machine.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
https://github.com/thin-edge/thin-edge.io/issues/1812

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [ x I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

As I discussed in the issue, the main goal here was to create a single source of truth for a couple of reasons:
1. The keys that are accepted by tedge config are generally not quite the same as the keys in tedge.toml due to the manual input required.
2. We had documentation for each key in both doc comments and descriptions.
3. The level of indirection required to obtain a configuration value makes the library unnecessarily complex for new users.

Annoyingly, to achieve this, I did have to use some significant, albeit declarative, macros. The results are, I believe, quite clear in the rustdoc output, and with the odd exception, are named completely intuitively. I've documented the macro syntax with relevant comments in the invocation of the macro.

### CODE REVIEW

File: crates/core/tedge/src/cli/connect/jwt_token.rs [hunk 1]
```diff
@@ -7,13 +7,14 @@ use rumqttc::MqttOptions;
 use rumqttc::Outgoing;
 use rumqttc::Packet;
 use rumqttc::QoS::AtLeastOnce;
+use tedge_config::IpAddress;
 
-pub(crate) fn get_connected_c8y_url(port: u16, host: String) -> Result<String, ConnectError> {
+pub(crate) fn get_connected_c8y_url(port: u16, host: IpAddress) -> Result<String, ConnectError> {
```
Comment (Bravo555) [inline crates/core/tedge/src/cli/connect/jwt_token.rs:12]: As above, I think that may be incorrect because hosts can be also DNS names, but I'm not sure

File: crates/core/tedge/src/cli/connect/common_mosquitto_config.rs [hunk 2]
```diff
@@ -135,15 +136,15 @@ impl CommonMosquittoConfig {
     pub fn with_external_opts(
         self,
         port: Option<u16>,
-        bind_address: Option<String>,
+        bind_address: Option<IpAddress>,
```
Comment (Bravo555) [inline crates/core/tedge/src/cli/connect/common_mosquitto_config.rs:139]: I think that may be incorrect because hosts can be also DNS names, but I'm not sure @didier-wenzek

File: crates/common/tedge_config/src/tedge_config_cli/new_tedge_config.rs [hunk 3]
```diff
@@ -0,0 +1,1412 @@
...
+use crate::ConnectUrl;
+use crate::IpAddress;
+use crate::TEdgeConfigLocation;
+use crate::TemplatesSet;
+
+use super::tedge_config_dto::TEdgeConfigDto;
+
+/// The configuration of thin-edge
+///
+/// This handles fetching values from [TedgeConfigDto](super::tedge_config_dto::TEdgeConfigDto) as well as from the system
```
Comment (Bravo555) [inline crates/common/tedge_config/src/tedge_config_cli/new_tedge_config.rs:33]: I see there is some hard wrapping for most doc comments, but in some places rows are very short and in some places it looks like they're very wide. So you could some of the comments to some arbitrary threshold, but I don't think either thin-edge or rustdoc mandate it, so it's just a suggesstion. I personally wrap comments to 80 columns and use Rewrap extension for VSCode for automatic wrapping.

**Review by Bravo555 — CHANGES_REQUESTED** [review]
I like how the new `tedge config list --doc` looks. However, some comments:

File: crates/common/tedge_config/src/tedge_config_cli/new_tedge_config.rs [hunk 4]
```diff
@@ -0,0 +1,1412 @@
...
+///
+/// // Getters for other settings will fail if the value is not setting
+/// assert!(config.mqtt_external_port().is_err());
+///
+/// // The defaults for some values are inferred from the config location
+/// assert!(config.device_key_path().starts_with(temp_dir.utf8_path()));
+///
+/// # Ok::<_, TEdgeConfigError>(())
+/// ```
+pub struct NewTEdgeConfig {
```
[the remark this answers is not in the document]
Comment (jarhodes314) [reply crates/common/tedge_config/src/tedge_config_cli/new_tedge_config.rs:56]: We could, although I'd prefer to keep the simple, unambiguous solution for now. I intend to update all the existing usage and completely replace `OldTEdgeConfig` with `NewTEdgeConfig` (and rename the struct and module to not contain new), albeit likely in a separate PR.

File: crates/common/tedge_config/src/lib.rs [hunk 5]
```diff
@@ -4,8 +4,9 @@ pub mod tedge_config_cli;
 pub use self::tedge_config_cli::config_setting::*;
 pub use self::tedge_config_cli::error::*;
 pub use self::tedge_config_cli::models::*;
+pub use self::tedge_config_cli::new_tedge_config::*;
+pub use self::tedge_config_cli::old_tedge_config::*;
```
[the remark this answers is not in the document]
Comment (jarhodes314) [reply crates/common/tedge_config/src/lib.rs:8]: I agree, although I think for this PR it's probably sensible to keep the major structural changes together, and deal with this naming issue separately.

File: crates/core/tedge/src/cli/connect/jwt_token.rs (hunk 1 above)
Comment (didier-wenzek) [reply crates/core/tedge/src/cli/connect/jwt_token.rs:12]: @Bravo555 is correct we need to be able to give a host name.

File: crates/core/tedge/src/cli/connect/common_mosquitto_config.rs (hunk 2 above)
Comment (didier-wenzek) [reply crates/core/tedge/src/cli/connect/common_mosquitto_config.rs:139]: For this specific case, I think `IpAddress` is more appropriate. A DNS name makes sense for `mqtt.client.host` but not the `mqtt.external.bind_address` that is the address bound by the local listener.

**Comment by Bravo555** [comment]
> If a user uses a key with lots of dots rather than the updated key, using tedge config will look something like this (I also renamed software.plugin.default -> software.default_plugin as the word order became particularly clunky after replacing the dot with an underscore):

So I have some concerns about simplifying TOML key names in order to accommodate env variable names. Sorry for not raising this earlier, but only now when implementing MQTT authentication, I see that I use up to 4 levels of nesting (`mqtt.client.auth.cert_file`) and I see that in this PR it was necessary to flatten out the TOML to be 2 levels deep.

So I wonder whether or not do we get kind of the worst of both worlds here. The advantage of TOML is that it can be deeply hierarchical,  contain objects, arrays, etc. but we restrict it two two levels only so that we can better support pretty looking environment variable names.

Have we considered the option of completely disallowing underscores (`_`) in our TOML keys? They could be replaced with `.` and become "dotted keys". Example from [TOML spec](https://toml.io/en/v1.0.0#keys):

```toml
name = "Orange"
physical.color = "orange"
physical.shape = "round"
site."google.com" = true
```

- closely related settings like `mqtt.external.bind_address` and `mqtt.external.bind_interface` would become `mqtt.external.bind.address` and `mqtt.external.bind.interface` respectively
- in some settings we could eliminate underscores altogether like `run.lock_files` -> `run.lockfiles`
- in some settings where we couldn't do the above we would introduce some unnecessary nesting, which would require some additional structs

This excess nesting sounds pretty painful, but I'd rather have that than only nesting 2 levels deep.

**Comment by jarhodes314** [comment]
> > If a user uses a key with lots of dots rather than the updated key, using tedge config will look something like this (I also renamed software.plugin.default -> software.default_plugin as the word order became particularly clunky after replacing the dot with an underscore):
> 
> So I have some concerns about simplifying TOML key names in order to accommodate env variable names. Sorry for not raising this earlier, but only now when implementing MQTT authentication, I see that I use up to 4 levels of nesting (`mqtt.client.auth.cert_file`) and I see that in this PR it was necessary to flatten out the TOML to be 2 levels deep.
> 
> So I wonder whether or not do we get kind of the worst of both worlds here. The advantage of TOML is that it can be deeply hierarchical, contain objects, arrays, etc. but we restrict it two two levels only so that we can better support pretty looking environment variable names.
> 
> Have we considered the option of completely disallowing underscores (`_`) in our TOML keys? They could be replaced with `.` and become "dotted keys". Example from [TOML spec](https://toml.io/en/v1.0.0#keys):
> 
> ```toml
> name = "Orange"
> physical.color = "orange"
> physical.shape = "round"
> site."google.com" = true
> ```
> 
>     * closely related settings like `mqtt.external.bind_address` and `mqtt.external.bind_interface` would become `mqtt.external.bind.address` and `mqtt.external.bind.interface` respectively
> 
>     * in some settings we could eliminate underscores altogether like `run.lock_files` -> `run.lockfiles`
> 
>     * in some settings where we couldn't do the above we would introduce some unnecessary nesting, which would require some additional structs
> 
> 
> This excess nesting sounds pretty painful, but I'd rather have that than only nesting 2 levels deep.

There's no technical reason why there needs to only be one dot to support environment variable names, it just made the logic for translating them to the correct key much simpler as it can be done without any knowledge of what the keys are. So long as the keys are unambiguous (e.g. we don't have keys like c8y.root.cert.path and c8y.root_cert_path meaning separate things), we can map the environment variables to variables successfully. The reason why the TOML is two levels deep here is because that is what is was. I was trying to (mostly) preserve the existing `tedge.toml` format as a source of truth at least for now to avoid the complexity of migrating the TOML to a new structure (because even if we want to rename a bunch of fields to make the names make more sense in a non-hierarchical structure, we can just tell serde to deal with it).

If we do change the structure, we absolutely need to get the naming somewhat self consistent. From what I can tell, `mqtt.port` is currently the counterpart to `mqtt.bind.address`. I think we should definitely keep using underscores to some degree.  Things like `c8y.root.cert.path` don't make any sense as a hierarchical structure. I agree it's a problem with mqtt, and most of the other keys are probably better as nested keys, but the impact will be pretty small.

FWIW, my (pretty recent, but maybe not since the last `toml_edit` update recent) experience with trying to turn "normal" TOML tables into dotted keys with `toml_edit` was awkward, and there were a bunch of issues with how it handled comments (which I should check for the existence of and raise issues for), although that probably doesn't have much impact on us. It does, however, have an impact if we wish to update the TOML in a format-preserving manner at some point, which could be more possible if we've already written manual migration logic for old TOML files.

File: crates/core/tedge/src/cli/connect/jwt_token.rs (hunk 1 above)
Comment (jarhodes314) [reply crates/core/tedge/src/cli/connect/jwt_token.rs:12]: If that's the case, the logic calling this that can only ever be an `IpAddress` is broken. I presume it's a case of us wanting to use the MQTT client address rather than the bind address? My gut feeling there is that the client address default should be the bind address (whereas the current default is just a copy of the bind address default)?

**Review by didier-wenzek — COMMENTED** [review]
I still need some time to have an overall views of the impacts of the proposed changes.

Looking things independently, I saw many interesting points compared to the old version. However, I still have a feeling that this if by far too complex for the task. The [`configuration_keys!`](https://github.com/thin-edge/thin-edge.io/pull/1916/files#diff-9891fe5d3e150caac7850fa4cba04aecbc0eb8e055a689632d1eea427cab393aR657) invocation  highlights how many weirdness we have introduced over time. Dealing with all of them is a "tour de force" and I admire you ingenuity to cope with this legacy. However, I wonder if wouldn't be simpler to give up maintaining backward compatibility on some of these points.

- Do we really need to have a `TedgeConfig` and `TedgeConfigDto`?
- Can this be simplified  once the `OldTedgeConfig` fully deprecated?

File: crates/common/tedge_config/src/tedge_config_cli/new_tedge_config.rs [hunk 6]
```diff
@@ -0,0 +1,1412 @@
...
+        self.stored
+            .mqtt()
+            .external_bind_address()
+            .unwrap_or_else(|| self.stored.mqtt().bind_address())
+    }
+
+    /// Getter for `firmware.child_update_timeout`
+    pub fn firmware_child_update_timeout(&self) -> Duration {
+        self.stored.firmware().child_update_timeout()
+    }
```
[the remark this answers is not in the document]
Comment (jarhodes314) [reply crates/common/tedge_config/src/tedge_config_cli/new_tedge_config.rs:119]: This is a special case because it felt sensible for any Rust code accessing this to access this as a `Duration`, but `tedge config` needs to read and write this as a number of seconds. So when we read the string value, we get the number of seconds, but when we read the typed value, we get a `Duration`.

File: crates/common/tedge_utils/src/file.rs
[the remark this answers is not in the document]
Comment (jarhodes314) [reply crates/common/tedge_utils/src/file.rs:1]: `whoami` doesn't seem to support getting group names, and those tests don't work on my machine unless I use an actual group name.

[the remark this answers is not in the document]
Comment (didier-wenzek) [reply crates/common/tedge_utils/src/file.rs:1]: > whoami doesn't seem to support getting group names, and those tests don't work on my machine unless I use an actual group name.

Okay. So let's keep these changes in your PR.

File: crates/core/tedge/src/cli/config/commands/list.rs [hunk 7]
```diff
@@ -53,8 +46,73 @@ fn print_config_list(
...
+        println!(
+            "{}  {}",
+            yansi::Paint::yellow(
+                key.pad_to_width_with_alignment(max_length, pad::Alignment::Right)
+            ),
+            yansi::Paint::default(docs).italic()
+        );
+
+        // TODO add a test to make sure people don't accidentally set the wrong meta name
+        if let Some(note) = ty.metas.get("note") {
```
[the remark this answers is not in the document]
Comment (jarhodes314) [reply crates/core/tedge/src/cli/config/commands/list.rs:81]: Doku has official support for examples, so we just use `#[doku(example="value")]`, but for notes we have to use metas (`#[doku(meta(note = "A note"))]` ) which is doku's mechanism for arbitrary structured metadata.

File: crates/common/tedge_config/src/tedge_config_cli/models/fakeable.rs
[the remark this answers is not in the document]
Comment (jarhodes314) [reply crates/common/tedge_config/src/tedge_config_cli/models/fakeable.rs:1]: That understanding is correct, I'll improve the docs

**Comment by jarhodes314** [comment]
> I still need some time to have an overall view of the impacts of the proposed changes.
> 
> Looking things independently, I saw many interesting points compared to the old version. However, I still have a feeling that this if by far too complex for the task. The [`configuration_keys!`](https://github.com/thin-edge/thin-edge.io/pull/1916/files#diff-9891fe5d3e150caac7850fa4cba04aecbc0eb8e055a689632d1eea427cab393aR657) invocation highlights how many weirdness we have introduced over time. Dealing with all of them is a "tour de force" and I admire you ingenuity to cope with this legacy. However, I wonder if it wouldn't be simpler to give up maintaining backward compatibility on some of these points.
> 
>     * Do we really need to have a `TedgeConfig` and `TedgeConfigDto`?
> 
>     * Can this be simplified  once the `OldTedgeConfig` fully deprecated?

There are two opposing forces at play here:
- We want to edit `tedge.toml` programmatically, and in doing so, get the benefits of using `serde` to its full potential, such as simple support for aliases and renaming fields. When we do that, we don't want to serialize default values in `tedge.toml`, it makes it harder to see what's configured and difficult/impossible to have logic deriving those defaults (such as deriving the mqtt client details from the bind details, or the device certificates from the configuration path).
- We don't want consumers of `tedge_config` to have to care that values are optional if they have defaults.

The first point requires we represent the absence of values, but the second requires we don't. Initially, I thought we could just have getters on `TEdgeConfigDto` to alleviate this, but the default certificate paths are derived from `TEdgeConfigLocation`, so anyone accessing those variables would have to keep that around and somewhat arbitrarily pass it in (although we may be able to use (something like) https://docs.rs/figment/latest/figment/value/magic/struct.RelativePathBuf.html to fix that).

One nice feature of the existing solution that I'm completely stuck on how I'd implement without code generation is the errors for when data isn't supplied. At the moment, we quote what configuration key isn't set, and this makes it really simple for consuming code to fail with a clear instructions on how to fix the problem. Of course you could specify manually in each of the relevant getters what the key name is, but then we're back to relying on someone to get the thing right in the first place.

**Comment by jarhodes314** [comment]
I've now updated the PR to include @Bravo555's changes. The behaviour should be identical to before, with the exception of the `mqtt.client_auth.cert_file`/`mqtt.client_auth.key_file` now including `_`s to match the dto (but the keys do work complete with the second dot, I've improved the logic to normalize the keys to allow for this). I don't think the added complexity of trying to update the toml manually (as opposed to modifying the dto and then serializing that and letting serde-derive do all the hard work) is worth it to save a few lines of pretty simple logic.

Now we have actually got an instance of a multi-level hierarchical configuration, it seems weird to limit this to this one configuration. I think it makes sense to update the toml to be more hierarchical, especially mqtt. This change probably doesn't belong in this PR, mainly because it doesn't need to and this PR is already rather large. We could introduce a version field to indicate whether the configuration is updated or not, and manually (or possibly automatically, like I do here for outdated key detection) modify a `toml::Value` to move e.g. `external_port` to `external.port`. If this is something we want to do, it would be worth disabling the "deprecated key" warnings until that change is completed as we would reverse many of the changes we are warning about.

On the subject of warnings, they're pretty coarse-grained at the moment, they don't have any knowledge of previous keys, so they will claim that any key with the right segments (e.g. `c8y_root.cert_path`) is deprecated (and point the user in the direction of the actual key). I don't think this is a huge issue, but there might be some unintended side effects that someone else can think of?

I've also attempted (not in anything I've pushed to here) to investigate what it would be like to replace this with a procedural macro that would replace most of `new_tedge_config.rs` and all of `tedge_config_dto.rs`, and I've got something that pretty much works. My proposed syntax is something like the following:

```rust
define_tedge_config! {
    mqtt: {
        #[tedge_config(default(function = "default_port"))]
        port: NonZeroU16,
        
        // I think this is a (slightly) useful feature for this one case.
        // If mqtt.port is set, but not mqtt.client_port, mqtt.client_port
        // should probably match it rather than defaulting to 1883.
        #[tedge_config(default(config_path = "mqtt.port"))]
        client_port: NonZeroU16,

        #[tedge_config(all_or_nothing)]
        client_auth: {
            cert_file: Utf8PathBuf,
            key_file: Utf8PathBuf,
        },
    },
   // ...
}

fn default_port() -> NonZeroU16 {
    1883.try_into().unwrap()
}
```

It still supports all the other attributes we're using from serde and doku, I omitted them here to focus on the new attributes. Using this would be something like:

```rust
let config: TEdgeConfig = ...;
assert_eq!(config.read().mqtt.port, default_port());
assert_eq!(config.read().mqtt.client_port, default_port());

match config.read().mqtt.client_auth {
     Ok(Some(auth)) => println!("{}, {}", auth.cert_file, auth.key_file),
     Ok(None) => println!("Client auth not set"),
     // It generates an appropriate error message if the config
     // is partially set
     Err(e) => println!("Invalid config: {e}"),
}
```

This solution has different trade offs. The parsing logic is mostly simpler to understand, and possible to unit test, but if the generated code is invalid, the error messages can be much more impenetrable. Usually with a declarative macro, rustc can point you at the bit of the macro that's failing, but emitting invalid syntax from a proc macro will generally give you no information at all as to where the error arose from. It would, however, simplify the code defining the configuration.

File: crates/common/tedge_config/src/tedge_config_cli/new_tedge_config.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (didier-wenzek) [reply crates/common/tedge_config/src/tedge_config_cli/new_tedge_config.rs:122]: I would prefer to address the issue at the `TEdgeConfigDto` level using `#[serde(deserialize_with = "to_duration", serialize_with = "from_duration")]`.

[the remark this answers is not in the document]
Comment (jarhodes314) [reply crates/common/tedge_config/src/tedge_config_cli/new_tedge_config.rs:122]: In which case we actually would need a wrapper type because it needs ToString and FromStr too for tedge config

**Comment by didier-wenzek** [comment]
superseded by https://github.com/thin-edge/thin-edge.io/pull/1936
